Refactor app component

germanferrero opened the pull request:
This big PR (sorry for that) is a step forward to remove redux and rx-js libraries from the LimeApp, as they usually represent a barrier for new developers, or libremesh developers coming from different backgrounds who want to collaborate with the LimeApp.
The core logic of the app which was previously managed by lime-plugin-core redux store + rx-js epics is now handled at an AppContextProvider following [React Context API](https://reactjs.org/docs/context.html).
Redux and rx-js is still being used by the majority of the plugins, we can progressively replace it one by one.
This change triggered a refactor of many parts of the application, that's why this PR is so huge.

In the way, I did other changes:
- Add ProtectedRoutes: plugins can configure that the user need to input the shared password to visit the plugin
- Handle session app-wise: these allow the user to authenticate once for every auth protected screen.
- Clean up the layout of the app.
- Clean up the app component code.
- Fix webpack config to let us use less-modules at /containers and /plugins
- Make storybook use the webpack config of the limeapp, so stories look the same as in the app.
- Remove not used code

germanferrero committed:
refactor(app): refactor the app component to be more clear

germanferrero committed:
refactor(app): remove empty stylesheet

germanferrero committed:
refactor(app): put banner inside content

germanferrero committed:
refactor(meta): remove not used constant and epics

germanferrero committed:
refactor(app): move store initialization to index.js

It's more easy to find that logic there

germanferrero committed:
refactor(ground-routing): remove unused ground routing code

germanferrero committed:
refactor(navigator): remove unused navigator component

germanferrero committed:
refactor(rx): remove empty style file

germanferrero committed:
improvement(less styles): config webpack to use less modules everywhere

Before, less modules could be used only at src/components

germanferrero committed:
chore(eslint): add eslint for react hooks

germanferrero committed:
improvement(less modules): fix less modules building in prod env

germanferrero committed:
refactor(app): replace redux + rx-js with Context and Hooks at the core

Move functionality from core plugin to appContext. Adds shared session between plugins. Adds
communityProtected routes, to ask for shared password before render. Simplify app component code

germanferrero committed:
revert(package.json): revert removal of .map files at build script

germanferrero committed:
improvement(stories): adapt stories after refactor

Also make storybook use limeapp webpack config

germanferrero committed:
fix(style): remove green button after recovering it

with the new way of importing skeletton, green button was recovered, but it wasnt intended

germanferrero committed:
fix(routes): community protected routes wait for connection

germanferrero committed:
fix(stories): recover necessary export at app.context

spiccinini reviewed (approved):
Great work!!!

nicopace commented on plugins/lime-plugin-admin/index.js:
```diff
@@ -15,5 +15,6 @@ export default {
 		constants
-	}
+	},
+	isCommunityProtected: true
```
would be good to describe the purpose of the default attributes and what the true/false value means.

nicopace commented on src/utils/app.context.js:
```diff
@@ -0,0 +1,169 @@
+
+export class AppContextProvider extends Component {
+
+	constructor(props) {
```
Would be good to know what explicit attributes could be passed to props... can be documented as a jsdoc for example

nicopace commented on plugins/lime-plugin-admin/src/adminApi.js:
```diff
@@ -1,19 +1,2 @@
-
-export const changeConfig = (api, sid, config) =>
-	api.call(sid, 'lime-utils', 'change_config', { name: config.hostname, ip: config.ip });
+export const changeConfig = (api, config) =>
```
if we are assuming that the context manages the authentification, these call should handle the unauthorization through an exception and a try/catch, with a global catch-all.
Responsability-wise I think it would be the best.

nicopace commented on plugins/lime-plugin-admin/src/adminPage.js:
```diff
@@ -35,12 +35,21 @@ const style = {
 
-export const Admin = ({ adminLogin, changeConfig, showNotification, selectedHost, nodeData, loading, authStatus }) => {
+export const Admin = ({ changeConfig, showNotification, nodeData, loading, redirect, error }) => {
+	const { nodeHostname, changeNode } = useAppContext();
```
it looks like useAppContext became a singleton on the system. would encourage to check if it is essential to be used in that way or a refactor is needed.

nicopace commented on plugins/lime-plugin-admin/src/adminPage.js:
```diff
@@ -83,42 +82,34 @@ export const Admin = ({ adminLogin, changeConfig, showNotification, selectedHost
 
 	return (
-		<div class="container" style={{ paddingTop: '100px' }}>
+		<div className="container container-padded">
```
replace className with class

nicopace replied:
this issue repeats in the whole PR

germanferrero replied:
will address in a different PR, since it's not a bug, and className is present at many files in develop too.

nicopace commented on plugins/lime-plugin-align/src/alignActions.js:
```diff
@@ -4,12 +4,13 @@ import {
 	SIGNAL_GET
 } from './alignConstants';
-import { store } from '../../../src/store';
+// import { store } from '../../../src/store';
```
remove

nicopace commented on plugins/lime-plugin-align/src/alignActions.js:
```diff
@@ -4,12 +4,13 @@ import {
-	if (iface === store.getState().align.currentReading.iface) {
-		return { type: 'no_iface' };
-	}
+	// TODO: FIX This store usage
```
fix before merge

germanferrero replied:
It ended up not being an issue, no fix needed, thanks!

nicopace commented on plugins/lime-plugin-align/src/alignActions.js:
```diff
@@ -19,9 +20,10 @@ export const changeInterface = (iface) => {
-	if (mac === store.getState().align.currentReading.mac) {
-		return { type: 'no_change' };
-	}
+	// TODO: FIX This store usage
```
fix before merging

nicopace commented on plugins/lime-plugin-ground-routing/src/groundRoutingReducer.js:
```diff
@@ -1,36 +0,0 @@
-import {
```
have you removed lime-plugin-ground-routing completely?

germanferrero replied:
I did a rollback of these change, and will open a separate issue for this soon

nicopace commented on src/components/app.js:
```diff
@@ -1,117 +1,60 @@
+
+const AppDefault = () => (
+	<AppContextProvider>
+		{/* TODO: Remove the dependency on redux store in the plugins */}
```
remove comment, or move it to an issue

germanferrero replied:
done!

germanferrero committed:
fix(align): revert align actions comments

germanferrero committed:
Revert "refactor(ground-routing): remove unused ground routing code"

This reverts commit c272b76064f65c345d162166b48ae3a23e341fe8.

germanferrero committed:
style(app): remove TODO comment

spiccinini commented:
Si corro   `npm run storybook` veo lo siguiente

![image](https://user-images.githubusercontent.com/500841/88621516-0dd57500-d077-11ea-83cf-2095b5098d8f.png)

germanferrero committed:
fix(align stories): fix store usage in align actions

An unnecesary usage of the store in align actions was triggering a weird bug in storybook.

germanferrero committed:
improvement(changenode): implement change node changing window.location

Before this, change node will only change the backend url. This change simplifies the code as it
doesnt have to handle corner cases anymore and prevent future bugs.

spiccinini commented:
:rocket:
